File: LeetCode/Explore/Queue & Stack/23. Merge k Sorted Lists.py
#Heap
#Time: O(nlogk)
#Space: O(k) 
class Solution:
    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        if not lists:
            return None
        
        heap = []
        
        #initialize
        for i in range(len(lists)):
            if lists[i] is not None:
                #adding i in the middle as a tie breaker
                #if the value is the same, heap cannot compare listNode
                #so we put the index in the middle to solve it
                heapq.heappush(heap,(lists[i].val, i, lists[i]))
        head = ListNode(0)
        curr = head
        #length = len(lists)
        i = 0
        #iterate
        while heap:
            #is we don't change the size of the heap,
            #it is faster using heapreplace
            #we only pop it when necessary
            val, i, node = heap[0]#heapq.heappop(heap)
            curr.next = ListNode(val)
            curr = curr.next
            node = node.next
            if node is not None:
                heapq.heapreplace(heap,(node.val, i, node))
            else:
                heapq.heappop(heap)


        return head.next

File: LeetCode/Problems/C++/23. Merge k Sorted Lists.cpp
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if(!lists.size()) return NULL;
        
        ListNode* head = new ListNode();
        ListNode* node = head;
        priority_queue<pair<int, ListNode*>> merge_q;
        for(int i=0;i<lists.size();i++)
            if(lists[i])
                merge_q.push({-lists[i]->val,lists[i]});
        
        while(!merge_q.empty()){
            ListNode* curr = merge_q.top().second;
            merge_q.pop();
            node->next = curr;
            node = node->next;
            if(curr->next)
                merge_q.push({-curr->next->val, curr->next});
        }
        
        
        return head->next;
    }
};